### ISSUE TITLE
fix(ml): refuse to overwrite an arm's fold artifacts without --force

### ISSUE THREAD

**Issue by LukeSantossz** [LukeSantossz/visiosoil-app#26]
> **Rescoped 2026-09-02, at the Developer's direction.** Four of this issue's
> five sub-problems died with SPEC 0042 — see the premise correction below, which
> is kept. What survives is sub-problem 5, and it is now the whole issue. The
> title and the acceptance criteria have been rewritten to say so, because they
> asked for `best_model.keras`, which SPEC 0042's nested selection **refuses by
> design**: the refit that produces the saved weights is passed `val_entries=None`
> deliberately, and a cross-phase checkpoint chosen on validation accuracy is the
> second, unaudited selection that protocol exists to prevent. Implementing the
> old criteria would have contradicted an approved record.
>
> ## What this issue is now
>
> **Re-running an arm silently overwrites the artifacts a published result names.**
> `train_fold` writes into `fold_directory(arm_dir, repeat, fold)` and calls
> `model.save(directory / CHECKPOINT_FILENAME)` with no existence check;
> `crossval.run_arm` walks every repeat and fold unconditionally. There is no
> `--force` anywhere in `ml/`.
>
> It matters more under k-fold than it did under a single run: a result now spans
> k x R directories, so a partial re-run leaves an arm whose folds came from two
> different trainings, with nothing in the artifacts saying so.
>
> ## Acceptance criteria, replacing those below
>
> - `run_arm` refuses to write into a fold directory that already holds
>   `predictions.json`, naming the directory and the flag that would allow it.
> - `--force` exists, is off by default, and is recorded in the fold's artifacts
>   when used, so a re-run is visible in the result rather than only in a shell
>   history.
> - A refusal names every fold directory that already exists, not only the first,
>   so an operator learns the whole cost of the decision in one message.
> - Refusing costs nothing when the arm has never run: a clean `arm_dir` proceeds.
>
> Everything below is the issue as filed and as previously corrected. It is kept
> because the audit trail is what makes the rescope checkable.

> **Premise correction, 2026-09-02.** SPEC 0042 and ADR 0020 (merged in #207) replaced the single-split training entry point with per-fold training under nested selection, and three of this issue's five sub-problems no longer describe `ml/src/train.py`.
>
> - **Sub-problem 3 is gone by deletion.** No `ModelCheckpoint` call and no `best_model` string survives anywhere in `ml/`. There is no dead checkpoint left to make live.
> - **Sub-problem 2 changed mechanism.** `_fit_two_phase` attaches `EarlyStopping(restore_best_weights=True)` only when a validation set is passed, which is the inner selection fits. The refit that produces the saved weights is passed `val_entries=None` deliberately: the epoch count was already chosen on the inner folds, and re-selecting on any held-out slice of the training side would be a second selection nobody audited.
> - **Proposed Solution items 1, 2 and 3 are therefore refused by design, not merely unimplemented.** A `ModelCheckpoint` monitoring `val_accuracy` across both phases is exactly the leak `assert_selection_is_nested` (`ml/src/train.py`) exists to refuse. The matching acceptance criteria — Phase 1 writing `best_model.keras`, `export.py` preferring it, and `metrics.json` carrying both phases' best validation accuracy — are moot.
> - **Sub-problem 5 survives, and is the whole of what remains.** `train_fold` writes into `fold_directory(arm_dir, repeat, fold)` and calls `model.save(directory / "model.keras")` with no existence check, and `crossval.run_arm` walks every repeat and fold unconditionally. There is still no `--force` anywhere in `ml/`, so re-running an arm silently overwrites the artifacts a published result names — which matters more under k-fold than it did under one run, because a result now spans k x R directories.
>
> Every line citation below points at the pre-#207 file and no longer resolves. The original text is left intact.

## Description

Training can persist a model worse than the one it actually produced. Phase 1 is never checkpointed, Phase 2's early stopping restores only Phase 2's best epoch, and the unconditional `model.save()` writes whatever Phase 2 ended with — even if every fine-tuning epoch was worse than Phase 1. The checkpoint that would prevent this is written and never read.

> **Revalidated 2026-08-22** against `main` at 4635804. Four of the five original sub-problems survive unchanged; sub-problem 4 is fixed. All line citations below are corrected — the originals, and the copies in the architecture docs, point at pre-SPEC-0032 line numbers.

## Context

Source: ML pipeline audit of 2026-05-03, Claude Code plus Codex adversarial review. Revalidated by the code audit of 2026-08-22.

The SPEC 0031 through 0035 wave hardened seeding, image verification, dataset manifests and the preprocessing contract, and touched none of the checkpoint-selection or artifact-overwrite logic.

This matters more now than when it was filed. Experiment E0 is a gate whose verdict is a comparison between runs; a pipeline that can silently persist a degraded model makes every such comparison unreliable, independently of the seeding work that SPEC 0032 completed.

## Current State

**1. Phase 1 has no checkpoint. `OPEN`**

`ml/src/train.py:192-200` — `callbacks_phase1` contains only `ReduceLROnPlateau`. The Phase 1 `fit` at `:202-208` passes that list. No `ModelCheckpoint`, no `EarlyStopping`, so no Phase 1 epoch is ever persisted or tracked.

**2. Phase 2 can silently discard Phase 1's best result. `OPEN`**

`ml/src/train.py:220-225` constructs `EarlyStopping(monitor="val_accuracy", restore_best_weights=True)`. Its restore scope is the Phase 2 `fit` at `:241-248`, a separate call whose best-tracking starts fresh. `ml/src/train.py:260-262` then saves unconditionally:

```python
keras_path = output_dir / "model.keras"
model.save(keras_path)
```

Nothing compares Phase 2's best validation accuracy against Phase 1's. `history1` and `history2` are concatenated for reporting at `:250-255` and never compared.

**3. `best_model.keras` is dead code. `OPEN`**

Written at `ml/src/train.py:218` and `:233-238` via `ModelCheckpoint(save_best_only=True)`. Never read: `ml/src/export.py:36-46` resolves `model.keras`, then `model.h5`, then raises. `evaluate.py` does not read it either. A grep for `best_model` across project-owned `ml/` hits only `train.py` and two architecture documents.

*Scope note:* this sub-problem is also claimed by work item B3 in `ml-implementation-map.md:333`. One owner — resolve it here or there, not both.

**4. Splits validation runs after the model build. `FIXED`**

`ml/src/train.py:136-139` validates splits; `build_model(cfg)` is not called until `:179`. Everything data-side precedes the build. Fixed at 75e26a3 on 2026-05-03 — the same day this issue was filed, so the claim was stale on arrival. SPEC 0032 (2d4e96b) additionally moved image verification ahead of the build.

**5. Re-running a version silently overwrites its artifacts. `OPEN`**

`ml/src/train.py:118-123` does `mkdir(parents=True, exist_ok=True)` then a plain truncating write of `config.json`. Same for `runtime.json` at `:129-130`, `model.keras` at `:260-261`, `history.json` at `:265-266`. The CLI at `:271-277` registers only `--version` and `--config`; there is no `--force` anywhere in `ml/`. SPEC 0032 added one more clobberable artifact without adding a guard.

No test in `ml/tests/` covers checkpoint selection or overwrite behaviour.

## Proposed Solution

Track the best model across both phases, and make the artifact directory append-only by default.

1. Give Phase 1 a `ModelCheckpoint` writing to the same `best_model.keras`, monitoring `val_accuracy` with `save_best_only=True`. Because both phases then write to one checkpoint under the same monitor, the file holds the best epoch across the whole run and the cross-phase comparison needs no separate bookkeeping.
2. Have `export.py` load `best_model.keras` when present, falling back to `model.keras` only if it is absent, and record which one it used in `spec.json` provenance.
3. Record both phases' best validation accuracy in `metrics.json`, so a Phase 2 that failed to improve is visible in the artifact rather than only in the logs.
4. Refuse to write into an existing version directory unless `--force` is passed.

Criteria for this shape: one checkpoint under one monitor is simpler than comparing two histories and cannot disagree with itself; making `export` prefer the checkpoint is what turns the existing dead code into the fix; and `--force` matches the fail-loud convention SPEC 0032 established.

## Acceptance Criteria

- [ ] Phase 1 writes to `best_model.keras` under the same monitor as Phase 2.
- [ ] `export.py` prefers `best_model.keras` and records which artifact it converted.
- [ ] A test drives a two-phase run in which every Phase 2 epoch is worse than Phase 1, and asserts the exported model is Phase 1's. It must fail on the current code.
- [ ] `metrics.json` carries both phases' best validation accuracy.
- [ ] Re-running an existing version without `--force` exits non-zero and writes nothing; a test covers it.
- [ ] The B3 double-booking of sub-problem 3 is resolved in one place.
